### DISCUSSION TITLE
same dataset， same code， different collection name， search result different， the distance different

### DISCUSSION THREAD

**Discussion by cxjcjj** [milvus-io/milvus#28292] [Q&A and General discussion]
def build_milvus_db(collection_name):
    pid2id, id2pid = get_pid2id()
    t1 = time.time()
    connections.connect(host=HOST, port=PORT)
    has = utility.has_collection(collection_name)
    if has:
        utility.drop_collection(collection_name)
    print(f"Does collection {collection_name} exist in Milvus: {has}")
    fields = [
        FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False),
        FieldSchema(name="embeddings", dtype=DataType.FLOAT_VECTOR, dim=256)
    ]
    
    schema = CollectionSchema(fields, f"{collection_name} is the simplest demo to recall en embeddings")
    en_milvus = Collection(collection_name, schema)
    
    for batch in tqdm(chunks(iter_file('embs.json'), size=10000)):
        batch_pid = []
        batch_vec = []
        for b in batch:
            batch_pid.append(pid2id[b[0]])
            batch_vec.append(b[1])
            
        res = en_milvus.insert([batch_pid, batch_vec])
        en_milvus.flush()

    print('inseart finish')
    
    index = {
        "index_type": "IVF_SQ8",
        "metric_type": "IP",
        "params": {"nlist": 4096},
    }
    en_milvus.create_index("embeddings", index)
    print(f'en milvus build index finish')
    t2 = time.time()
    print('time:', t2 - t1)

def data_recall(evaluate_file, out_file, collection_name):
    pid2id, id2pid = get_pid2id()
    
    connections.connect(host=HOST, port=PORT)
    en_milvus = Collection(collection_name)
    en_milvus.load()
    
    pid2data = get_data('data.json')

    t1 = time.time()
    out_res = open(out_file, 'w', encoding='utf-8')

    search_params = {
        "metric_type": "IP", 
        "params": {"nprobe": 120}
    }

    for sub_batch_vec_id in tqdm(chunks(ids, 100)):
        sub_batch_str = ', '.join(sub_batch_vec_id)
        id_res = en_milvus.query(expr=f'id in [ {sub_batch_str} ]', output_fields=["embeddings"])
        
        search_ids = []
        search_emb = []
        for item in id_res:
            id = item['id']
            emb = item['embeddings']
            search_ids.append(id)
            search_emb.append(emb)
        
        result = en_milvus.search(data=search_emb, anns_field='embeddings', param=search_params, limit=10000)

![image](https://github.com/milvus-io/milvus/assets/20812193/61dde1ff-4db7-4f33-8396-a4b4fc9e5087)
![image](https://github.com/milvus-io/milvus/assets/20812193/2704dc27-a7a3-4be3-9f68-ab2bb8c271e6)
The order of the returned results is consistent, but there are differences in distance, most importantly when the topk expands to a certain number, there is a significant difference

**Comment by yhmo** [comment]
In "IVF_SQ8" index, vectors are quantized(compressed), the distance is a bit different to the real distance.
If the collection_A has finished index, and collection_B has not finished index, then their distance result is different.

**Reply by yhmo** [reply]
Change the index type to IVF_FLAT or FLAT, you will see equal distances.

**Reply by yhmo** [reply]
https://milvus.io/docs/index.md#IVF_SQ8
"**This index type can convert each FLOAT (4 bytes) to UINT8 (1 byte) by performing Scalar Quantization (SQ). This reduces disk, CPU, and GPU memory consumption by 70–75%.**"

**Reply by cxjcjj** [reply]
collectionA and B have fineshed index，but their distance result is different. My question is why the distance are different in the same situation.

**Reply by cxjcjj** [reply]
I accept that the distance is a bit different to the real distance. but I don't understand why repeating the same operation resulted in inconsistent results

**Reply by cxjcjj** [reply]
What is the method to ensure consistent recall results for the same data and operation every time by using "IVF_SQ8" index

**Reply by yhmo** [reply]
how many vectors in the 'embs.json'?

**Reply by cxjcjj** [reply]
1. 500w+ vectors
2. create_index() after insert() 
3. There is a delay in create_index() ? code finished, and there's no problem with the graphical interface either. 
![image](https://github.com/milvus-io/milvus/assets/20812193/239dedc7-cb15-48cc-83d6-e7d77ebefeac)

**Reply by yhmo** [reply]
```
for batch in tqdm(chunks(iter_file('embs.json'), size=10000)):
    batch_pid = []
    batch_vec = []
    for b in batch:
        batch_pid.append(pid2id[b[0]])
        batch_vec.append(b[1])
        
    res = en_milvus.insert([batch_pid, batch_vec])
    en_milvus.flush()
```

**Change this to the below and try again:**

```
for batch in tqdm(chunks(iter_file('embs.json'), size=10000)):
    batch_pid = []
    batch_vec = []
    for b in batch:
        batch_pid.append(pid2id[b[0]])
        batch_vec.append(b[1])
        
    res = en_milvus.insert([batch_pid, batch_vec])

en_milvus.flush()
```

**Reply by yhmo** [reply]
Background of milvus data management. This is important to understand how it does ANN search. And sometimes the results might be affected by the data management.

Note: **we never say it returns the same result for two collections with the same data, the distances might be different because they might have different segments and in different states of compaction process. but once they reach the same state, the results will be the same.**

The background:
- Milvus manages data by segments, ideally, each segment size is near to 512MB. But in runtime, there might be small segments generated, so, milvus will combine small segments into large segments.
- Milvus builds index for each segment, each segment has an independent index. when search(), each index returns topk results and these results are combined to be a final topk result

**Reply by cxjcjj** [reply]
Is there any way to ensure that the status is consistent?

**Reply by yhmo** [reply]
Each time you call flush(), it will force the datanode to generate a new segment.
You insert entities batch by batch, each batch is 10000 rows, that is to say, the collection will get 500+ small segments.
The datanode combines the small segments to large segments. For 256 dim vectors, each final segment contains 520000+ vectors. Finally, there will be 10 segments in each collection.
Since you call flush() too meny times, the datanode is busy to combine segments, the indexnode is also busy to build index for each new segment. When you call search(), the states of the two collections are different.

**Reply by yhmo** [reply]
Move en_milvus.flush() outside of the loop.

**Reply by cxjcjj** [reply]
I tried to move en_milvus.flush() outside of the loop, but the result was still different

**Reply by yhmo** [reply]
The IVF_SQ8 index quantizes each vector, the quantized result is affected by all the vectors. Different vectors or different numbers of vectors will get different quantized results.
Since milvus has its internal flush logic, with the same data, different collections might get different numbers of segments. Some segments contain 300000 vectors, some segments contain 500000 vectors. So, the quantized result could be different.
With different quantized results, the distance is also different. That is why a vector could return different distances in different collections.

If you are an expert on the FAISS SQ8 index, then you will understand why this happens.

Because of the quantize behavior, the recall rate of IVF_SQ8 is also worse than IVF_FLAT and HNSW. Usually, its recall rate is less than 95%, while the IVF_FALT/HNSQ can easily reach 99%. If you want to get a high recall rate, use IVF_FLAT/HNSW instead.
